src/lib.rs

```diff
@@ -218,10 +212,166 @@ use srml_support::traits::Currency;
 use srml_support::{decl_event, decl_module, decl_storage, dispatch, StorageValue};
 use system::ensure_signed;
 
-pub trait MembershipRegistry<T: system::Trait> {
-    fn ensure_member(who: T::AccountId) -> Result<(), &'static str>;
+use rstd::collections::btree_map::BTreeMap;
+
+/// Constant values
+/// Later add to st
+///
+
+/// Represents a user in this forum.
+pub trait ForumUser<AccountId> {
+    
+    /// Identifier of user 
+    fn id(&self) -> &AccountId;
+
+    // In the future one could add things like 
+    // - updating post count of a user
+    // - updating status (e.g. hero, new, etc.)
+    //
+
+}
+
+/// Represents a regsitry of `ForumUser` instances.
+pub trait ForumUserRegistry<AccountId> {
+
+    type ForumUser: ForumUser<AccountId>;
+    
+    fn get_forum_user(&self, id:AccountId) -> Option<&Self::ForumUser>;
+
+}
+
+/// Convenient composite time stamp 
+struct BlockchainTimestamp<BlockNumber, Moment> {
+    block : BlockNumber,
+    time: Moment
+}
+
+//#[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug))]
+/// Represents a moderation outcome applied to a post or a thread. 
+struct ModerationAction<BlockNumber, Moment, AccountId> {
+
+    /// When action occured.
+    occured_at: BlockchainTimestamp<BlockNumber, Moment>,
+
+    /// Account forum sudo which acted.
+    forum_sudo_account_id: AccountId,
+
+    /// Moderation rationale
+    rationale: Vec<u8>
+
 }
 
+/// Represents a revision of the text of a Post
+struct PostTextEdit<BlockNumber, Moment> {
+
+    /// What this edit occured.
+    occured_at: BlockchainTimestamp<BlockNumber, Moment>,
+
+    /// New text
+    new_text: Vec<u8>
+}
+
+/// Represents a thread post
+type PostId = u64;
+struct Post<BlockNumber, Moment, AccountId> {
+
+    /// Post identifier
+    id: PostId,
+
+    /// Id of thread to which this post corresponds.
+    thread_id: ThreadId,
+
+    // Position of post in thread.
+    position: u64,
+
+    /// Initial text of post
+    initial_text: Vec<u8>,
+
+    /// Possible moderation of this post
+    moderation : Option<ModerationAction<BlockNumber, Moment, AccountId>>,
+    
+    /// Edits of post ordered chronologically by edit time.
+    post_text_edits: Vec<PostTextEdit<BlockNumber, Moment>>,
+
+    /// When post was submitted.
+    created_at : BlockchainTimestamp<BlockNumber, Moment>,
+    
+    /// Author of post.
+    author_id : AccountId
+
+}
+
+/// Represents a thread
+type ThreadId = u64;
+struct Thread<BlockNumber, Moment, AccountId> {
+
+    /// Thread identifier
+    id : ThreadId,
+
+    /// Title
+    title : Vec<u8>,
+
+    /// Category in which this thread lives
+    category: CategoryId,
```

Comment:
Inconsistency in naming:
```suggestion
    category_id: CategoryId,
```

Reply:
Fixed.